srtcore/channel.cpp

```diff
@@ -126,53 +113,298 @@ CChannel::~CChannel()
+            m_sockstate = 1;
+        }
+    }
+}
+
+int EventRunner::socketReady() const
+{
+    return m_sockstate;
+}
+
+int EventRunner::signalReading(PipeSignal val) const
+{
+    // Setting this value as an intermediate passing storage.
+    // This will be picked up by EventRunner::poll as this
+    // shall cause immediate exit in WSAWaitForMultipleEvents().
+    m_state = val;
+    if (WSASetEvent(m_Event[WE_TRIGGER]))
+        return 0;
+    return WSAGetLastError();
+}
+
+PipeSignal EventRunner::clearSignalReading()
+{
+    WSAResetEvent(m_Event[WE_TRIGGER]);
+
+    // The state that was rewritten once poll() returned.
+    PipeSignal laststate = m_permstate;
+    m_state = PSG_NONE;
+    m_permstate = PSG_NONE;
+    return laststate;
+}
+
+// POSIX/fallback version. This applies to all systems
+// that can use ::select with both pipes and sockets.
+#else
+void EventRunner::init(int socket)
+{
+    m_fdSocket = socket;
+
+    if (::pipe(m_fdTrigger) == -1)
+        throw CUDTException(MJ_SETUP, MN_NONE, NET_ERROR);
+
+    int opts = ::fcntl(m_fdTrigger[PIPE_IN], F_GETFL);
+    if (opts == -1 || ::fcntl(m_fdTrigger[PIPE_IN], F_SETFL, opts | O_NONBLOCK) == -1)
+        throw CUDTException(MJ_SETUP, MN_NONE, NET_ERROR);
+}
+
+EventRunner::~EventRunner()
+{
+    ::close(m_fdTrigger[PIPE_IN]);
+    ::close(m_fdTrigger[PIPE_OUT]);
+}
+
+void EventRunner::poll(int64_t timeout_us)
+{
+    FD_ZERO(&in_set);
+    FD_ZERO(&err_set);
+
+    // Add the UDP socket
+    FD_SET(m_fdSocket, &in_set);
+    FD_SET(m_fdSocket, &err_set);
+    FD_SET(m_fdTrigger[PIPE_IN], &in_set);
+    int nsockets = std::max(m_fdSocket, m_fdTrigger[PIPE_IN]) + 1;
+    timeval tv;
+    tv.tv_usec = timeout_us;
+    tv.tv_sec = 0;
+
+    if (::select(nsockets, &in_set, NULL, &err_set, &tv) == -1)
+    {
+        LOGC(mglog.Error, log << "EventRunner::poll: IPE: select:"
+                << SysStrError(NET_ERROR));
+        // But continue the work.
+    }
+}
+
+int EventRunner::socketReady() const
+{
+    return FD_ISSET(m_fdSocket, &in_set) + 2*FD_ISSET(m_fdSocket, &err_set);
+}
+
+int EventRunner::signalReading(PipeSignal val) const
+{
+    char data[2] = { val, 0 }; // could be 0, but this sounds bad.
+    return ::write(m_fdTrigger[PIPE_OUT], data, 1);
+}
+
+PipeSignal EventRunner::clearSignalReading()
+{
+    int value = -1;
+
+    if (!FD_ISSET(m_fdTrigger[PIPE_IN], &in_set))
+        return PSG_NONE;
+
+    char onebyte[2];
+
+    // Ignore errors. This is in order to purge
+    // the pipe of the data that were stuffed into it just
+    // to cause a signal on the select that is waiting for
+    // read-readiness of the UDP socket.
+    int ret = ::read(m_fdTrigger[PIPE_IN], onebyte, 1);
+    if (ret == -1)
+        return PSG_NONE;
+
+    value = onebyte[0];
+    return PipeSignal(value);
+}
+
+#endif
+
 void CChannel::open(const sockaddr* addr)
 {
-   // construct an socket
-   m_iSocket = ::socket(m_iIPversion, SOCK_DGRAM, 0);
+    // construct an socket
```

Comment:
an socket -> a socket

Reply:
Probably I wanted to construct something else previously :)